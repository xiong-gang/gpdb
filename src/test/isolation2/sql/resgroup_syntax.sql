-- ----------------------------------------------------------------------
-- Test: assign/alter a role to a resource group
-- ----------------------------------------------------------------------

DROP ROLE IF EXISTS rg_test_role;

-- positive
CREATE ROLE rg_test_role;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
CREATE ROLE rg_test_role_super SUPERUSER;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role_super';

ALTER ROLE rg_test_role RESOURCE GROUP none;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
ALTER ROLE rg_test_role_super RESOURCE GROUP none;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role_super';

ALTER ROLE rg_test_role RESOURCE GROUP default_group;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
ALTER ROLE rg_test_role_super RESOURCE GROUP admin_group;
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role_super';


-- negative
ALTER ROLE rg_test_role RESOURCE GROUP non_exist_group;
ALTER ROLE rg_test_role RESOURCE GROUP admin_group;

CREATE ROLE rg_test_role1 RESOURCE GROUP non_exist_group;
-- nonsuper user should not be assigned to admin group
CREATE ROLE rg_test_role1 RESOURCE GROUP admin_group;

-- cleanup
DROP ROLE rg_test_role;
DROP ROLE rg_test_role_super;

-- ----------------------------------------------------------------------
-- Test: create/drop a resource group
-- ----------------------------------------------------------------------

--start_ignore
DROP RESOURCE GROUP rg_test_group;
--end_ignore

SELECT * FROM gp_toolkit.gp_resgroup_config;

-- negative

-- can't create the reserved resource groups
CREATE RESOURCE GROUP default_group WITH (cpu_rate_limit=10, memory_limit=10);
CREATE RESOURCE GROUP admin_group WITH (cpu_rate_limit=10, memory_limit=10);
CREATE RESOURCE GROUP none WITH (cpu_rate_limit=10, memory_limit=10);
-- multiple resource groups can't share the same name
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10);
DROP RESOURCE GROUP rg_test_group;
-- must specify both memory_limit and cpu_rate_limit
CREATE RESOURCE GROUP rg_test_group WITH (memory_limit=10);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10);
-- can't specify the resource limit type multiple times
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=5, memory_limit=5, concurrency=1);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=5, memory_limit=5, cpu_rate_limit=5);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=5, memory_limit=5, memory_limit=5);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=5, memory_limit=5, memory_shared_quota=70, memory_shared_quota=80);
-- can't drop non-exist resource group
DROP RESOURCE GROUP non_exist_group;
-- can't drop reserved resource groups
DROP RESOURCE GROUP default_group;
DROP RESOURCE GROUP admin_group;
DROP RESOURCE GROUP none;

-- positive
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10);
SELECT groupname,concurrency,proposed_concurrency,cpu_rate_limit,memory_limit,proposed_memory_limit,memory_shared_quota,memory_spill_ratio FROM gp_toolkit.gp_resgroup_config WHERE groupname='rg_test_group';
DROP RESOURCE GROUP rg_test_group;
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=10, memory_limit=10, memory_shared_quota=70, memory_spill_ratio=30);
SELECT groupname,concurrency,proposed_concurrency,cpu_rate_limit,memory_limit,proposed_memory_limit,memory_shared_quota,memory_spill_ratio FROM gp_toolkit.gp_resgroup_config WHERE groupname='rg_test_group';
DROP RESOURCE GROUP rg_test_group;

-- ----------------------------------------------------------------------
-- Test: boundary check in create resource group syntax
-- ----------------------------------------------------------------------

-- negative: cpu_rate_limit & memory_limit should be in [1, 100]
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=61, memory_limit=10);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=61);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=0, memory_limit=10);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=0);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=0.9);
-- negative: concurrency should be in [1, max_connections]
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=-1, cpu_rate_limit=10, memory_limit=10);
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=0, cpu_rate_limit=10, memory_limit=10);
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=26, cpu_rate_limit=10, memory_limit=10);
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=26, cpu_rate_limit=10, memory_limit=10);
-- negative: (memory_shared_quota + memory_spill_ratio) should be in [1, 100]
--           memory_spill_ratio > 0
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=10, memory_spill_ratio=0);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=10, memory_spill_ratio=-1);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=-1, memory_spill_ratio=10);
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=50, memory_spill_ratio=51);

-- positive: cpu_rate_limit & memory_limit should be in [1, 100]
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=60, memory_limit=10);
DROP RESOURCE GROUP rg_test_group;
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=60);
DROP RESOURCE GROUP rg_test_group;
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=1, memory_limit=10);
DROP RESOURCE GROUP rg_test_group;
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=1);
DROP RESOURCE GROUP rg_test_group;
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=60.9);
DROP RESOURCE GROUP rg_test_group;
-- positive: concurrency should be in [1, max_connections]
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=10, memory_limit=10);
DROP RESOURCE GROUP rg_test_group;
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=25, cpu_rate_limit=10, memory_limit=10);
DROP RESOURCE GROUP rg_test_group;
CREATE RESOURCE GROUP rg1_test_group WITH (concurrency=1, cpu_rate_limit=10, memory_limit=10);
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=50, memory_limit=50);
DROP RESOURCE GROUP rg1_test_group;
DROP RESOURCE GROUP rg2_test_group;
CREATE RESOURCE GROUP rg1_test_group WITH (concurrency=1, cpu_rate_limit=20, memory_limit=20);
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=40, memory_limit=40);
DROP RESOURCE GROUP rg1_test_group;
DROP RESOURCE GROUP rg2_test_group;
CREATE RESOURCE GROUP rg1_test_group WITH (concurrency=1, cpu_rate_limit=30, memory_limit=30);
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=30, memory_limit=30);
DROP RESOURCE GROUP rg1_test_group;
DROP RESOURCE GROUP rg2_test_group;
-- positive: (memory_shared_quota + memory_spill_ratio) should be in [1, 100]
--           memory_spill_ratio > 0
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=0, memory_spill_ratio=1);
DROP RESOURCE GROUP rg_test_group;
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=50, memory_spill_ratio=50);
DROP RESOURCE GROUP rg_test_group;
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=0, memory_spill_ratio=100);
DROP RESOURCE GROUP rg_test_group;
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=99, memory_spill_ratio=1);
DROP RESOURCE GROUP rg_test_group;

-- ----------------------------------------------------------------------
-- Test: alter a resource group
-- ----------------------------------------------------------------------
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=5, memory_limit=5);

-- ALTER RESOURCE GROUP SET CONCURRENCY N
-- negative: concurrency should be in [1, max_connections]
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 0;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -1;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 26;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -0.5;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 0.5;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY a;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 'abc';
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY '1';
-- positive: concurrency should be in [1, max_connections]
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 1;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 2;
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 25;

-- ALTER RESOURCE GROUP SET CPU_RATE_LIMIT VALUE
-- negative: cpu_rate_limit & memory_limit should be in [1, 100]
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT -0.1;
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT -1;
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0;
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.7;
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 1.7;
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 61;
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT a;
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 'abc';
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 20%;
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.2%;
-- positive: cpu_rate_limit & memory_limit should be in [1, 100]
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 1;
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 2;
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 60;
DROP RESOURCE GROUP rg_test_group;
-- positive: total cpu_rate_limit & memory_limit should be in [1, 100]
CREATE RESOURCE GROUP rg1_test_group WITH (cpu_rate_limit=10, memory_limit=10);
CREATE RESOURCE GROUP rg2_test_group WITH (cpu_rate_limit=10, memory_limit=10);
ALTER RESOURCE GROUP rg1_test_group SET CPU_RATE_LIMIT 50;
ALTER RESOURCE GROUP rg1_test_group SET CPU_RATE_LIMIT 40;
ALTER RESOURCE GROUP rg2_test_group SET CPU_RATE_LIMIT 20;
ALTER RESOURCE GROUP rg1_test_group SET CPU_RATE_LIMIT 30;
ALTER RESOURCE GROUP rg2_test_group SET CPU_RATE_LIMIT 30;
DROP RESOURCE GROUP rg1_test_group;
DROP RESOURCE GROUP rg2_test_group;

