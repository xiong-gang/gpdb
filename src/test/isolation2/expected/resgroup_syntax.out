-- ----------------------------------------------------------------------
-- Test: assign/alter a role to a resource group
-- ----------------------------------------------------------------------

DROP ROLE IF EXISTS rg_test_role;
DROP

-- positive
CREATE ROLE rg_test_role;
CREATE
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
rolresgroup
-----------
6437       
(1 row)
CREATE ROLE rg_test_role_super SUPERUSER;
CREATE
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role_super';
rolresgroup
-----------
6438       
(1 row)

ALTER ROLE rg_test_role RESOURCE GROUP none;
ALTER
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
rolresgroup
-----------
6437       
(1 row)
ALTER ROLE rg_test_role_super RESOURCE GROUP none;
ALTER
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role_super';
rolresgroup
-----------
6438       
(1 row)

ALTER ROLE rg_test_role RESOURCE GROUP default_group;
ALTER
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role';
rolresgroup
-----------
6437       
(1 row)
ALTER ROLE rg_test_role_super RESOURCE GROUP admin_group;
ALTER
SELECT rolresgroup FROM pg_authid WHERE rolname = 'rg_test_role_super';
rolresgroup
-----------
6438       
(1 row)


-- negative
ALTER ROLE rg_test_role RESOURCE GROUP non_exist_group;
ERROR:  resource group "non_exist_group" does not exist
ALTER ROLE rg_test_role RESOURCE GROUP admin_group;
ERROR:  only superuser can be assigned to admin resgroup

CREATE ROLE rg_test_role1 RESOURCE GROUP non_exist_group;
ERROR:  resource group "non_exist_group" does not exist
-- nonsuper user should not be assigned to admin group
CREATE ROLE rg_test_role1 RESOURCE GROUP admin_group;
ERROR:  only superuser can be assigned to admin resgroup

-- cleanup
DROP ROLE rg_test_role;
DROP
DROP ROLE rg_test_role_super;
DROP

-- ----------------------------------------------------------------------
-- Test: create/drop a resource group
-- ----------------------------------------------------------------------

--start_ignore
DROP RESOURCE GROUP rg_test_group;
ERROR:  resource group "rg_test_group" does not exist
--end_ignore

SELECT * FROM gp_toolkit.gp_resgroup_config;
groupid|groupname    |concurrency|proposed_concurrency|cpu_rate_limit|memory_limit|proposed_memory_limit|memory_shared_quota|proposed_memory_shared_quota|memory_spill_ratio|proposed_memory_spill_ratio
-------+-------------+-----------+--------------------+--------------+------------+---------------------+-------------------+----------------------------+------------------+---------------------------
6437   |default_group|20         |20                  |30            |30          |30                   |50                 |50                          |20                |20                         
6438   |admin_group  |2          |2                   |10            |10          |10                   |50                 |50                          |20                |20                         
(2 rows)

-- negative

-- can't create the reserved resource groups
CREATE RESOURCE GROUP default_group WITH (cpu_rate_limit=10, memory_limit=10);
ERROR:  resource group "default_group" already exists
CREATE RESOURCE GROUP admin_group WITH (cpu_rate_limit=10, memory_limit=10);
ERROR:  resource group "admin_group" already exists
CREATE RESOURCE GROUP none WITH (cpu_rate_limit=10, memory_limit=10);
ERROR:  resource group name "none" is reserved
-- multiple resource groups can't share the same name
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10);
CREATE
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10);
ERROR:  resource group "rg_test_group" already exists
DROP RESOURCE GROUP rg_test_group;
DROP
-- must specify both memory_limit and cpu_rate_limit
CREATE RESOURCE GROUP rg_test_group WITH (memory_limit=10);
ERROR:  must specify both memory_limit and cpu_rate_limit
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10);
ERROR:  must specify both memory_limit and cpu_rate_limit
-- can't specify the resource limit type multiple times
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=5, memory_limit=5, concurrency=1);
ERROR:  Find duplicate resoure group resource type: concurrency
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=5, memory_limit=5, cpu_rate_limit=5);
ERROR:  Find duplicate resoure group resource type: cpu_rate_limit
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=5, memory_limit=5, memory_limit=5);
ERROR:  Find duplicate resoure group resource type: memory_limit
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=5, memory_limit=5, memory_shared_quota=70, memory_shared_quota=80);
ERROR:  Find duplicate resoure group resource type: memory_shared_quota
-- can't drop non-exist resource group
DROP RESOURCE GROUP non_exist_group;
ERROR:  resource group "non_exist_group" does not exist
-- can't drop reserved resource groups
DROP RESOURCE GROUP default_group;
ERROR:  cannot drop default resource group "default_group"
DROP RESOURCE GROUP admin_group;
ERROR:  cannot drop default resource group "admin_group"
DROP RESOURCE GROUP none;
ERROR:  resource group "none" does not exist

-- positive
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10);
CREATE
SELECT groupname,concurrency,proposed_concurrency,cpu_rate_limit,memory_limit,proposed_memory_limit,memory_shared_quota,memory_spill_ratio FROM gp_toolkit.gp_resgroup_config WHERE groupname='rg_test_group';
groupname    |concurrency|proposed_concurrency|cpu_rate_limit|memory_limit|proposed_memory_limit|memory_shared_quota|memory_spill_ratio
-------------+-----------+--------------------+--------------+------------+---------------------+-------------------+------------------
rg_test_group|20         |20                  |10            |10          |10                   |20                 |20                
(1 row)
DROP RESOURCE GROUP rg_test_group;
DROP
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=10, memory_limit=10, memory_shared_quota=70, memory_spill_ratio=30);
CREATE
SELECT groupname,concurrency,proposed_concurrency,cpu_rate_limit,memory_limit,proposed_memory_limit,memory_shared_quota,memory_spill_ratio FROM gp_toolkit.gp_resgroup_config WHERE groupname='rg_test_group';
groupname    |concurrency|proposed_concurrency|cpu_rate_limit|memory_limit|proposed_memory_limit|memory_shared_quota|memory_spill_ratio
-------------+-----------+--------------------+--------------+------------+---------------------+-------------------+------------------
rg_test_group|1          |1                   |10            |10          |10                   |70                 |30                
(1 row)
DROP RESOURCE GROUP rg_test_group;
DROP

-- ----------------------------------------------------------------------
-- Test: boundary check in create resource group syntax
-- ----------------------------------------------------------------------

-- negative: cpu_rate_limit & memory_limit should be in [1, 100]
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=61, memory_limit=10);
ERROR:  total cpu_rate_limit exceeded the limit of 100
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=61);
ERROR:  total memory_limit exceeded the limit of 100
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=0, memory_limit=10);
ERROR:  cpu_rate_limit range is [1, 100]
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=0);
ERROR:  memory_limit range is [1, 100]
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=0.9);
ERROR:  memory_limit range is [1, 100]
-- negative: concurrency should be in [1, max_connections]
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=-1, cpu_rate_limit=10, memory_limit=10);
ERROR:  concurrency range is [1, 'max_connections']
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=0, cpu_rate_limit=10, memory_limit=10);
ERROR:  concurrency range is [1, 'max_connections']
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=26, cpu_rate_limit=10, memory_limit=10);
ERROR:  concurrency range is [1, 'max_connections']
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=26, cpu_rate_limit=10, memory_limit=10);
ERROR:  concurrency range is [1, 'max_connections']
-- negative: (memory_shared_quota + memory_spill_ratio) should be in [1, 100]
--           memory_spill_ratio > 0
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=10, memory_spill_ratio=0);
ERROR:  memory_spill_ratio range is [1, 100]
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=10, memory_spill_ratio=-1);
ERROR:  memory_spill_ratio range is [1, 100]
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=-1, memory_spill_ratio=10);
ERROR:  memory_shared_quota range is [0, 100]
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=50, memory_spill_ratio=51);
ERROR:  The sum of memory_shared_quota (50) and memory_spill_ratio (51) exceeds 100

-- positive: cpu_rate_limit & memory_limit should be in [1, 100]
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=60, memory_limit=10);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=60);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=1, memory_limit=10);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=1);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=60.9);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP
-- positive: concurrency should be in [1, max_connections]
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=1, cpu_rate_limit=10, memory_limit=10);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP
CREATE RESOURCE GROUP rg_test_group WITH (concurrency=25, cpu_rate_limit=10, memory_limit=10);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP
CREATE RESOURCE GROUP rg1_test_group WITH (concurrency=1, cpu_rate_limit=10, memory_limit=10);
CREATE
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=50, memory_limit=50);
CREATE
DROP RESOURCE GROUP rg1_test_group;
DROP
DROP RESOURCE GROUP rg2_test_group;
DROP
CREATE RESOURCE GROUP rg1_test_group WITH (concurrency=1, cpu_rate_limit=20, memory_limit=20);
CREATE
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=40, memory_limit=40);
CREATE
DROP RESOURCE GROUP rg1_test_group;
DROP
DROP RESOURCE GROUP rg2_test_group;
DROP
CREATE RESOURCE GROUP rg1_test_group WITH (concurrency=1, cpu_rate_limit=30, memory_limit=30);
CREATE
CREATE RESOURCE GROUP rg2_test_group WITH (concurrency=1, cpu_rate_limit=30, memory_limit=30);
CREATE
DROP RESOURCE GROUP rg1_test_group;
DROP
DROP RESOURCE GROUP rg2_test_group;
DROP
-- positive: (memory_shared_quota + memory_spill_ratio) should be in [1, 100]
--           memory_spill_ratio > 0
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=0, memory_spill_ratio=1);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=50, memory_spill_ratio=50);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=0, memory_spill_ratio=100);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=10, memory_limit=10, memory_shared_quota=99, memory_spill_ratio=1);
CREATE
DROP RESOURCE GROUP rg_test_group;
DROP

-- ----------------------------------------------------------------------
-- Test: alter a resource group
-- ----------------------------------------------------------------------
CREATE RESOURCE GROUP rg_test_group WITH (cpu_rate_limit=5, memory_limit=5);
CREATE

-- ALTER RESOURCE GROUP SET CONCURRENCY N
-- negative: concurrency should be in [1, max_connections]
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 0;
ERROR:  concurrency limit cannot be less than 1
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -1;
ERROR:  concurrency limit cannot be less than 1
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 26;
ERROR:  concurrency limit cannot be greater than 'max_connections'
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -0.5;
ERROR:  syntax error at or near "0.5"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY -0.5;
                                                            ^
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 0.5;
ERROR:  syntax error at or near "0.5"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 0.5;
                                                           ^
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY a;
ERROR:  syntax error at or near "a"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY a;
                                                           ^
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 'abc';
ERROR:  syntax error at or near "'abc'"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 'abc';
                                                           ^
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY '1';
ERROR:  syntax error at or near "'1'"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY '1';
                                                           ^
-- positive: concurrency should be in [1, max_connections]
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 1;
ALTER
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 2;
ALTER
ALTER RESOURCE GROUP rg_test_group SET CONCURRENCY 25;
ALTER

-- ALTER RESOURCE GROUP SET CPU_RATE_LIMIT VALUE
-- negative: cpu_rate_limit & memory_limit should be in [1, 100]
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT -0.1;
ERROR:  syntax error at or near "0.1"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT -0.1;
                                                               ^
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT -1;
ERROR:  cpu rate limit must be greater than 1
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0;
ERROR:  cpu rate limit must be greater than 1
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.7;
ERROR:  syntax error at or near "0.7"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.7;
                                                              ^
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 1.7;
ERROR:  syntax error at or near "1.7"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 1.7;
                                                              ^
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 61;
ERROR:  total cpu_rate_limit exceeded the limit of 100
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT a;
ERROR:  syntax error at or near "a"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT a;
                                                              ^
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 'abc';
ERROR:  syntax error at or near "'abc'"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 'abc';
                                                              ^
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 20%;
ERROR:  syntax error at or near "%"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 20%;
                                                                ^
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.2%;
ERROR:  syntax error at or near "0.2"
LINE 1: ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 0.2%;
                                                              ^
-- positive: cpu_rate_limit & memory_limit should be in [1, 100]
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 1;
ALTER
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 2;
ALTER
ALTER RESOURCE GROUP rg_test_group SET CPU_RATE_LIMIT 60;
ALTER
DROP RESOURCE GROUP rg_test_group;
DROP
-- positive: total cpu_rate_limit & memory_limit should be in [1, 100]
CREATE RESOURCE GROUP rg1_test_group WITH (cpu_rate_limit=10, memory_limit=10);
CREATE
CREATE RESOURCE GROUP rg2_test_group WITH (cpu_rate_limit=10, memory_limit=10);
CREATE
ALTER RESOURCE GROUP rg1_test_group SET CPU_RATE_LIMIT 50;
ALTER
ALTER RESOURCE GROUP rg1_test_group SET CPU_RATE_LIMIT 40;
ALTER
ALTER RESOURCE GROUP rg2_test_group SET CPU_RATE_LIMIT 20;
ALTER
ALTER RESOURCE GROUP rg1_test_group SET CPU_RATE_LIMIT 30;
ALTER
ALTER RESOURCE GROUP rg2_test_group SET CPU_RATE_LIMIT 30;
ALTER
DROP RESOURCE GROUP rg1_test_group;
DROP
DROP RESOURCE GROUP rg2_test_group;
DROP

